src/io/flutter/sdk/FlutterPluginsLibraryManager.java

```diff
@@ -0,0 +1,235 @@
+/*
+ * Copyright 2017 The Chromium Authors. All rights reserved.
+ * Use of this source code is governed by a BSD-style license that can be
+ * found in the LICENSE file.
+ */
+package io.flutter.sdk;
+
+import com.intellij.ProjectTopics;
+import com.intellij.openapi.application.ApplicationManager;
+import com.intellij.openapi.application.ModalityState;
+import com.intellij.openapi.application.WriteAction;
+import com.intellij.openapi.module.Module;
+import com.intellij.openapi.module.ModuleManager;
+import com.intellij.openapi.project.DumbService;
+import com.intellij.openapi.project.Project;
+import com.intellij.openapi.roots.*;
+import com.intellij.openapi.roots.impl.libraries.LibraryEx;
+import com.intellij.openapi.roots.impl.libraries.LibraryTableBase;
+import com.intellij.openapi.roots.impl.libraries.ProjectLibraryTable;
+import com.intellij.openapi.roots.libraries.Library;
+import com.intellij.openapi.roots.libraries.LibraryTable;
+import com.intellij.openapi.roots.libraries.LibraryTablesRegistrar;
+import com.intellij.openapi.util.text.StringUtil;
+import com.intellij.openapi.vfs.*;
+import com.jetbrains.lang.dart.util.DotPackagesFileUtil;
+import io.flutter.pub.PubRoot;
+import io.flutter.utils.FlutterModuleUtils;
+import org.jetbrains.annotations.NotNull;
+
+import java.util.*;
+import java.util.concurrent.atomic.AtomicBoolean;
+
+import static com.jetbrains.lang.dart.util.PubspecYamlUtil.PUBSPEC_YAML;
+
+public class FlutterPluginsLibraryManager {
+  private final Project project;
+
+  public FlutterPluginsLibraryManager(@NotNull Project project) {
+    this.project = project;
+  }
+
+  public void startWatching() {
+    VirtualFileManager.getInstance().addVirtualFileListener(new VirtualFileContentsChangedAdapter() {
+      @Override
+      protected void onFileChange(@NotNull VirtualFile file) {
+        fileChanged(project, file);
+      }
+
+      @Override
+      protected void onBeforeFileChange(@NotNull VirtualFile file) {
+      }
+    }, project);
+
+    project.getMessageBus().connect().subscribe(ProjectTopics.PROJECT_ROOTS, new ModuleRootListener() {
+      @Override
+      public void rootsChanged(ModuleRootEvent event) {
+        scheduleUpdate();
+      }
+    });
+
+    scheduleUpdate();
+  }
+
+  private void fileChanged(@NotNull final Project project, @NotNull final VirtualFile file) {
+    if (!DotPackagesFileUtil.DOT_PACKAGES.equals(file.getName())) return;
+    if (LocalFileSystem.getInstance() != file.getFileSystem() && !ApplicationManager.getApplication().isUnitTestMode()) return;
+
+    final VirtualFile parent = file.getParent();
+    final VirtualFile pubspec = parent == null ? null : parent.findChild(PUBSPEC_YAML);
+
+    if (pubspec != null) {
+      scheduleUpdate();
+    }
+  }
+
+  private final AtomicBoolean isUpdating = new AtomicBoolean(false);
```

Comment:
Do we have a convention about placement of declarations? I always put fields at the top of a class, and tend to expect to see them there. Actually, I usually reformat + reorganize code, which puts things in order.

Reply:
I'll sort this nearer to the top -